### ISSUE TITLE
[BUG] reliable_unordered is not reliable

### ISSUE THREAD

**Issue by CosineP** [amethyst/laminar#174]
See this example: https://gist.github.com/CosineP/52e5068375bd1ea11374fc4dac34c1d3

If packets are lost, they never make it to receiver (for example, i'm using `tc qdisc change dev eth0 root netem loss 10% 25%`)

this applies:
- in the wild (not a TC fluke)
- on linux and macos

**Comment by fhaynes** [comment]
I was able to replicate this issue on macOS using dummynet:

```
dnctl list
00001: unlimited    0 ms   50 sl.plr 0.100000 1 queues (1 buckets) droptail
```

plr is the packet loss rate (0.1 is 10%), and is reliably dropping 10% of packets, when laminar should be resending

**Comment by TimonPost** [comment]
I fixed the bug and code for that can be found over here: <https://github.com/TimonPost/laminar/tree/bug_fix>.

If you run `cargo test acking_500_packets_with_packet_drop` you will run the test in `acknowlegement.rs`. 

Some results I get back:
```
100 sent packets: 17 dropped should be 25
200 sent packets: 42 dropped should be 50
400 sent packets: 92 dropped should be 100
800 sent packets: 192 dropped should be 200
```

First, there were 0 packets dropped, I checked out the dropping code and discovered a bug, when I fixed the bug I saw, as you can see above, always 8 packets missing. This has to do something some not correct parameter.

**Comment by TimonPost** [comment]
I did the following changes: 

_acknowlegement.rs_
```rust
pub fn process_incoming(&mut self, incoming_seq: u16) {
    self.their_acks.ack(incoming_seq);
    let dropped_packets = self.waiting_packets.ack(incoming_seq, self.bit_mask());
    unsafe { dropped_packet_count += dropped_packets.len() as u32; }
    println!("count of current dropped packets: {}", unsafe { dropped_packet_count });
    self.dropped_packets.extend(dropped_packets.into_iter().map(|(_, p)| p));
}
```

results I got back when running the test from the above comment:

```
count: 0 <-=
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   = first 24 are not dropped
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0   =
count: 0 <-=
count: 1 <-=
count: 1   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 1 <-=
count: 2 <-=
count: 2   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 2 <-=
count: 3 <-=
count: 3   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 3 <-=
count: 4 <-=
count: 4  = 3 times there won't happen any thing, 4th time a drop occurs.
count: 4 <-=
count: 5 <-=
count: 5  = 3 times there won't happen any thing, 4th time a drop occurs.
count: 5 <-=
count: 6 <-=
count: 6   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 6 <-=
count: 7 <-=
count: 7   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 7 <-=
count: 8 <-=
count: 8   = 3 times there won't happen any thing, 4th time a drop occurs.
count: 8 <-=
```

A noticeable thing here is that the first 24 packets won't do anything to the dropped packet count. 
- Why are those first 24 packets resulting in zero dropped packets, and why that many?

**Comment by CosineP** [comment]
does it have to do with this somehow? https://github.com/TimonPost/laminar/blob/bug_fix/src/infrastructure/local_ack.rs#L44 im trying to wrap my head around these bitfields and failing completely.... (why were they chosen over a HashMap in the first place?)

**Comment by CosineP** [comment]
got it!!!!

**Comment by CosineP** [comment]
Ok actually fixed it this time

**Comment by TimonPost** [comment]
As you probably don't know we use an acknowledgment way described by gaffer, his blog about it can be found over here: <https://gafferongames.com/post/reliability_ordering_and_congestion_avoidance_over_udp/>

This way, as he describes it was already implemented by a library called: gaffer_udp: <https://github.com/acmcarther/gaffer_udp/>, he said that this acknowledgment was working pretty well and it was covered well with unit tests. Because of that almost all of the acking code was copy paste into laminar. 

However, I just read this on his blog post:

_gaffer blogs_
```
In our reliability system, we never resend a packet with a given sequence number. We sequence n exactly once, then we send n+1, n+2 and so on. We never stop and resend packet n if it was lost, we leave it up to the application to compose a new packet containing the data that was lost, if necessary, and this packet gets sent with a new sequence number.
```

But we are currently storing the packets without headers in the dropped queue. When those dropped packets are resent they don't have any header nor sequence number attached, those packets are just the raw bytes of the payload. This, at least for that I now know of,  causes all dropped packets not to be accepted when they are resent. 
Those dropped packets have to get a new sequence id and will have to be processed. 

_currently this is done_
```
 for payload in dropped_packets {
                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
 }
```

This issue should be solved too, which will happen after merging #175
